src/Libraries/Microsoft.Extensions.ObjectPool.DependencyInjection/ObjectPoolServiceCollectionExtensions.cs

```diff
@@ -0,0 +1,96 @@
+        Action<PoolOptions>? configure = null)
+        where TService : class
+    {
+        return services.AddPooledInternal<TService, TService>(configure);
+    }
+
+    /// <summary>
+    /// Adds an <see cref="ObjectPool{TService}"/> and let DI return scoped instances of TService.
+    /// </summary>
+    /// <param name="services">The <see cref="IServiceCollection"/> to add to.</param>
+    /// <param name="configure">Configuration of the pool.</param>
+    /// <typeparam name="TService">The type of objects to pool.</typeparam>
+    /// <typeparam name="TImplementation">The type of the implementation to use.</typeparam>
+    /// <returns>Provided service collection.</returns>
+    /// <exception cref="ArgumentNullException"><paramref name="services"/> is <see langword="null"/>.</exception>
+    /// <remarks>
+    /// The default capacity is 1024.
+    /// The pooled type instances are obtainable by resolving <see cref="ObjectPool{TService}"/> from the DI container.
+    /// </remarks>
+    public static IServiceCollection AddPooled<TService, [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] TImplementation>(
+        this IServiceCollection services,
+        Action<PoolOptions>? configure = null)
+        where TService : class
+        where TImplementation : class, TService
+    {
+        return services.AddPooledInternal<TService, TImplementation>(configure);
+    }
+
+    /// <summary>
+    /// Registers an action used to configure the <see cref="PoolOptions"/> of a typed pool.
+    /// </summary>
+    /// <typeparam name="TService">The type of objects to pool.</typeparam>
+    /// <param name="services">The <see cref="IServiceCollection"/> to add the services to.</param>
+    /// <param name="configure">The action used to configure the options.</param>
+    /// <returns>The <see cref="IServiceCollection"/> so that additional calls can be chained.</returns>
+    public static IServiceCollection ConfigurePool<TService>(this IServiceCollection services, Action<PoolOptions> configure)
+        where TService : class
+    {
+        return services.Configure<PoolOptions>(typeof(TService).FullName, configure);
+    }
+
+    private static IServiceCollection AddPooledInternal<TService, [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] TImplementation>(
+        this IServiceCollection services,
+        Action<PoolOptions>? configure)
+        where TService : class
+        where TImplementation : class, TService
+    {
+        _ = Throw.IfNull(services);
+
+        if (configure != null)
+        {
+            // Register a PoolOption instance specific to the type
+            _ = services.ConfigurePool<TService>(configure);
+        }
+
+        return services
+            .AddSingleton<ObjectPool<TService>>(provider =>
+            {
+                var options = provider.GetService<IOptionsFactory<PoolOptions>>()?.Create(typeof(TService).FullName!) ?? new PoolOptions();
```

Comment:
How does this differ from `provider.GetRequiredService<IOptions<PoolOptions>>`?

Reply:
`Required` would fail if none is registered. But more importantly it has to be a different value per name (the pool type) which `GetRequiredService` doesn't provide.